### ISSUE TITLE
ProgressBar ETA for an IterableDataset

### ISSUE THREAD

**Issue by g-karthik** [pytorch/ignite#1263]
## ❓ Questions/Help/Support

We have a piece of code which uses `ignite` successfully and was originally written for a regular `TensorDataset` (adapted from [this](https://github.com/huggingface/transfer-learning-conv-ai/blob/master/train.py)).

But then we adapted it for an `IterableDataset` because we wanted to train on a much larger dataset (on the order of GBs). How does the `ProgressBar` estimate this ETA for an `IterableDataset`?

If there is any working sample code you can point me to on using `IterableDataset` instances successfully with ignite, that would be great!

@vfdev-5

**Comment by vfdev-5** [comment]
> How does the ProgressBar estimate this ETA for an IterableDataset?

@g-karthik engine can estimate the epoch length using the first epoch and then tqdm total is set according to that : 
https://github.com/pytorch/ignite/blob/d8a54f6aa230f52b8f77162b8dcdbe9baa7f6c35/ignite/contrib/handlers/tqdm_logger.py#L48-L51

Here is an example of how it would work
```python
# ignite 0.4.1
import time

from torch.utils.data import DataLoader, IterableDataset
from ignite.engine import Engine
from ignite.contrib.handlers import ProgressBar


class MyIterableDataset(IterableDataset):
    def __init__(self):
        super(MyIterableDataset).__init__()
        self.start = 0
        self.end = 23

    def __iter__(self):
        return iter(range(self.start, self.end))


num_workers = 0
ds = MyIterableDataset()
data_loader = DataLoader(ds, num_workers=num_workers)

def foo(e, b):
    time.sleep(0.2)

engine = Engine(foo)

ProgressBar().attach(engine)
engine.run(data_loader, epoch_length=None, max_epochs=5)
```

**Comment by g-karthik** [comment]
> engine can estimate the epoch length using the first epoch and then tqdm total is set according to that

@vfdev-5 but does this mean that the ETA reported for the first epoch is likely to be unrealistic? i.e., only from the second epoch the ETAs will be accurate?

**Comment by vfdev-5** [comment]
@g-karthik normally, during the first epoch you shouldn't have any ETA reported but something like that : 
```
Epoch [1/5]: [8/?] 0%   [....] [00:04<00:00]
```
and only from the second epoch tqdm can estimate the time as it has now the epoch length:
```
Epoch [2/5]: [8/23] 35% [00:01<00:03]
```

**Comment by g-karthik** [comment]
@vfdev-5 I kicked off a 1 epoch multi-node multi-GPU training job over the weekend and it finished. It was showing me an ETA though.

At the end of the job, it shows something like this:

``
Epoch: [138269/138269] 100%|##########, loss=0.394 [45:37:57<00:00]
``

During the start of the job, it was showing something like `[00:00<45:30:16]`, so I assumed the ETA was ~45 hours.

From what you're saying then, it sounds like I am doing something incorrectly? Also, the 138269 above seems to be iterations, not epochs.

**Comment by vfdev-5** [comment]
@g-karthik yes, I agree that looks a bit weird. I think I already saw previously something like that : 
```
Epoch: [138269/138269]
```
This is shown when you have `max_epochs=1`.  I agree that it would be nice to fix this visualization issue.

> During the start of the job, it was showing something like [00:00<45:30:16]

This is however very strange...

**Comment by vfdev-5** [comment]
I think if you are using DDP approach, it would be better to take `num_samples` as `len(dataloader.sampler)` as the dataset is split to your N processes ?

**Comment by g-karthik** [comment]
@vfdev-5 yes I am using DDP, but I don't have an explicit `sampler` associated with my `DataLoader`.

As for why compute and set iterations to all-reduce MIN of samples, it's because if we don't, then our `trainer` will attempt forward-backward and loss computation (on all processes, including the min one) even after the min process is done with its data, and we don't want that happening right? Because then the min process would fail/cause the overall train job to terminate?

Even if the overall `trainer` doesn't terminate and training continues on the remaining processes, if our code uses `world_size` and `world_rank` like variables for some logic, then these variables need to be updated immediately after the min process is done.

**Comment by vfdev-5** [comment]
> As for why compute and set iterations to all-reduce MIN of samples, it's because if we don't, then our trainer will attempt forward-backward and loss computation (on all processes, including the min one) even after the min process is done with its data, and we don't want that happening right? Because then the min process would fail/cause the overall train job to terminate?

Maybe, I'm missing some context here. Probably, I understand what you mean: each process loads its data which can have different number of samples. Then you would like to make all processes aware of the min number of samples. And this defines your epoch. Otherwise you can have a divergence between processes which could lead to a crash of hanging...


Anyway, I think it would make sense to file an issue about the display of progressbar as 
```
Epoch: [138269/138269] 
```
when `max_epochs=1` as "Epoch" should be replaced by "Iterations". @g-karthik can you send an issue for that please ? Thanks 

Have we clarified the question about ETA ?

**Comment by g-karthik** [comment]
@vfdev-5 yeah the ETA part is fine, closing this issue!
